opendir.c

```diff
@@ -47,143 +47,294 @@ struct watchman_dir_handle {
-  elem = pathcopy + 1;
-  while (elem) {
-    int next_fd;
+  canon = w_win_unc_to_utf8(long_buf, long_len);
+  if (!canon) {
+    return false;
+  }
 
-    // if elem -> "foo/bar" we want:
-    // elem -> "foo", next -> "bar"
-    // if there is no slash, say if elem -> "bar":
-    // elem -> "bar", next -> NULL
-    next = strchr(elem, '/');
-    if (next) {
-      *next = 0;
-      next++;
+  path_base = w_basename(path);
+  canon_base = w_basename(canon);
+  if (!path_base || !canon_base) {
+    result = false;
+  } else {
+    result = strcmp(path_base, canon_base) == 0;
+    if (!result) {
+      w_log(W_LOG_ERR,
+            "is_basename_canonical_case(%s): basename=%s != canonical "
+            "%s basename of %s\n",
+            path, path_base, canon, canon_base);
     }
+  }
+  free(canon);
 
-    // Ensure that we're CLOEXEC on all intermediate components of
-    // path, but use the flags the caller passed to us when we reach
-    // the leaf of the tree
-    next_fd = openat(fd, elem, (next ? O_CLOEXEC : flags)|O_NOFOLLOW);
-    if (next_fd == -1) {
-      err = errno;
-      close(fd);
-      free(pathcopy);
-      errno = err;
+  return result;
+#else
+  unused_parameter(path);
+  return true;
+#endif
+}
+
+/* Extract the canonical path of an open file descriptor and store
+ * it into the provided buffer.
+ * Unlike w_realpath, which will return an allocated buffer of the correct
+ * size, this will indicate EOVERFLOW if the buffer is too small.
+ * For our purposes this is fine: if the canonical name is larger than
+ * the input buffer it means that it does not match the path that we
+ * wanted to open; we don't care what the actual canonical path really is.
+ */
+static int realpath_fd(int fd, char *canon, size_t canon_size) {
+#if defined(F_GETPATH)
+  unused_parameter(canon_size);
+  return fcntl(fd, F_GETPATH, canon);
+#elif defined(__linux__)
+  char procpath[1024];
+  int len;
+
+  snprintf(procpath, sizeof(procpath), "/proc/%d/fd/%d", getpid(), fd);
+  len = readlink(procpath, canon, canon_size);
+  if (len > 0) {
+    canon[len] = 0;
+    return 0;
+  }
+  if (errno == ENOENT) {
+    // procfs is likely not mounted, let caller fall back to realpath()
+    errno = ENOSYS;
+  }
+  return -1;
+#elif defined(_WIN32)
+  HANDLE h = (HANDLE)_get_osfhandle(fd);
+  WCHAR final_buf[WATCHMAN_NAME_MAX];
+  DWORD len, err;
+  DWORD nchars = sizeof(final_buf) / sizeof(WCHAR);
+
+  if (h == INVALID_HANDLE_VALUE) {
+    return -1;
+  }
+
+  len = GetFinalPathNameByHandleW(h, final_buf, nchars,
+                                  FILE_NAME_NORMALIZED | VOLUME_NAME_DOS);
+  err = GetLastError();
+  if (len >= nchars) {
+    errno = EOVERFLOW;
+    return -1;
+  }
+  if (len > 0) {
+    char *utf8 = w_win_unc_to_utf8(final_buf, len);
+    if (!utf8) {
       return -1;
     }
 
-    // walk down to next level
-    elem = next;
-    close(fd);
-    fd = next_fd;
+    uint32_t utf_len = u32_strlen(utf8);
```

Comment:
Just a note: `u32_strlen()` is a terrible name for this function, because there is such a thing as UTF-32, and it's not clear from the name that this isn't returning the length of the UTF-8 string when encoded in UTF-32.

(Reading the implementation makes it clear it's actually `uint32_t`.)